lib/src/back_end/solution.dart

```diff
@@ -85,11 +85,18 @@ class Solution implements Comparable<Solution> {
     var piece = _state.firstUnsolved();
     if (piece == null) return const [];
 
-    return [
-      // All pieces support a default state.
-      Solution(root, pageWidth, _state.cloneWith(piece, State.initial)),
+    var states = [
+      if (piece.pinnedState case var state?)
+        state
+      else ...[
+        // All pieces support a default state.
```

Comment:
I think it would be OK to ship this solution, but it feels a little off to me, and I think this comment highlights why. Is it fair to say that this comment should read "All _non-pinned pieces_ support a default state"?

It's also interesting that we _aren't_ reading `piece.defaultState` - instead we bake in the assumption that all non-pinned pieces have a default state of `State.initial`. The coupling seems tighter than it needs to be - and I think `Piece` could own more of this knowledge.

What would you think of changing `Piece.states` to be the API used here, and subclasses override something else like `stateOptions`?

```dart
class Piece {
  List<State> get states => _pinnedState ?? [State.initial, ...stateOptions];

  List<State> get stateOptions;

  List<State>? _pinnedState;

  void pin(State state) {
    _pinnedState = [state];
  }
}
```

Reply:
Maybe this is along the same lines as what Nate is talking about, but it does seem a little confusing to me how pinned states and `states` itself is separate in `BlockPiece`. It makes sense to me that a pinned state would be a part of `states` like in Nate's example.

Reply:
Yeah, I agree this API is getting hairy. Having every Piece implicitly support `State.initial` (whose meaning is somewhat piece dependent) feels kind of hacky too. It avoids having to copy/paste `State.initial` (or something similar) in every piece subclass, but I'm not sure that's really that useful.

I took a stab at doing something similar to what you suggested and I think it looks better. PTAL.